IsExplicitMatch correctly overriden for CompositeFilters. In a Compos…

stffabi opened the pull request:
…iteFilter the IsExplicitMatch function must always be correctly combined by calling the sub-filters IsExplicitMatch function. Fixes #1141

rprouse commented on src/NUnitFramework/framework/Internal/Filters/CompositeFilter.cs:
```diff
@@ -64,6 +64,24 @@ public void Add(ITestFilter filter)
+        /// Checks whether the CompositeFilter is matched by a test.
+        /// </summary>
+        /// <param name="test">The test to be matched</param>
+        public abstract override bool Pass(ITest test);
```
I assume that you added these overrides to ensure they are implemented in all of the child classes. Correct?

stffabi replied:
Yes, exactly.

CharliePoole replied:
I find the override a bit confusing in reading the code. I would definitely add it if this were a class intended to be extended by users, but it's purely internal.

stffabi replied:
I think that's a matter of preference. I personally prefer it even in purely internal code to force implementers to override methods that doesn't make sense how they are implemented in the base class.

But I can drop that from the commit, that's not a big deal :smile:.

CharliePoole replied:
Except it's almost certain that all the derived classes we will ever have
are already implemented!

But it certainly does no harm.

On Mon, Apr 11, 2016 at 2:28 AM, stffabi <notifications@github.com> wrote:

> In src/NUnitFramework/framework/Internal/Filters/CompositeFilter.cs
> <https://github.com/nunit/nunit/pull/1400#discussion_r59176879>:
>
> > @@ -64,6 +64,24 @@ public void Add(ITestFilter filter)
> >          public IList<ITestFilter> Filters { get; private set; }
> >
> >          /// <summary>
> > +        /// Checks whether the CompositeFilter is matched by a test.
> > +        /// </summary>
> > +        /// <param name="test">The test to be matched</param>
> > +        public abstract override bool Pass(ITest test);
>
> I think that's a matter of preference. I personally prefer it even in
> purely internal code to force implementers to override methods that doesn't
> make sense how they are implemented in the base class.
>
> But I can drop that from the commit, that's not a big deal [image:
> :smile:].
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/nunit/nunit/pull/1400/files/9ca0604beef21f693409647b3389283c4333caf8#r59176879>
>

CharliePoole replied:
Except it's almost certain that all the classes we will ever have

In src/NUnitFramework/framework/Internal/Filters/CompositeFilter.cs
<https://github.com/nunit/nunit/pull/1400#discussion_r59176879>:

> @@ -64,6 +64,24 @@ public void Add(ITestFilter filter)
>          public IList<ITestFilter> Filters { get; private set; }
>
>          /// <summary>
> +        /// Checks whether the CompositeFilter is matched by a test.
> +        /// </summary>
> +        /// <param name="test">The test to be matched</param>
> +        public abstract override bool Pass(ITest test);

I think that's a matter of preference. I personally prefer it even in
purely internal code to force implementers to override methods that doesn't
make sense how they are implemented in the base class.

But I can drop that from the commit, that's not a big deal [image: :smile:].

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub
<https://github.com/nunit/nunit/pull/1400/files/9ca0604beef21f693409647b3389283c4333caf8#r59176879>

rprouse commented:
The code looks good, but we need tests for this. If you look in the framework tests under `Internal\Filters`, there are `AndFilterTests` and `OrFilterTests`. These have `ExplicitMatchTests` which clearly are not working correctly. Can you determine what is wrong with these tests and fix them? If you can, I would actually like to see passing and failing tests in both for Match, Pass and ExplicitMatch. I think that is likely the problem, the positive tests are all correct, but we never test for failed matches.

I would recommend taking a TDD approach to the tests. Comment out your fixes, add/fix the tests to get them failing, then uncomment your code to prove that your tests and your fixes are correct. Whenever I work on any issue in NUnit, I try to always start with the tests because it is too easy to write tests after the fact that pass, but do not properly test the fix, like this one. :smile:

stffabi commented:
Thanks for pointing out where the tests are, I'll take a look into it.

stffabi commented:
The `ExplicitMatchTests` are correct, the problem is that for the specific test-case there's no difference in the result as long as the sub-filter doesn't override `IsExcplicitMatch`.

I've added a test-case with the `NotFilter`, which overrides the `IsExcplicitMatch` function. Furthermore I've added some generic `CombinationTest`, these first test the combination of the sub-filters in the AndFilter/OrFilter and second also checks that the expected "Match"-Function gets called and not a wrong one.

CharliePoole commented on src/NUnitFramework/tests/Internal/Filters/AndFilterTests.cs:
```diff
@@ -71,6 +71,74 @@ public void ExplicitMatchTest()
+            Assert.AreEqual(expectedResult, calculatedResult);
+        }
+
+        private bool Match(
```
Having a test method with the same name as one of the underlying TestFilter methods is a bit confusing.

CharliePoole commented on src/NUnitFramework/tests/Internal/Filters/AndFilterTests.cs:
```diff
@@ -71,6 +71,74 @@ public void ExplicitMatchTest()
+                case StrictIdFilterForTests.EqualValueFunction.Pass:
+                    return filter.Pass(_dummyFixture);
+                default:
+                    throw new AssertionException(
```
Don't throw the AssertionException, rather, use Assert.Fail as we tell users to do. :-) This will stop working as soon as we change how failures are reported. Alternatively, if you want to convey that this is an error in how the test was written rather than a true failure, throw something else, like an ArgumentException.

CharliePoole commented on src/NUnitFramework/tests/Internal/Filters/OrFilterTests.cs:
```diff
@@ -76,6 +76,58 @@ public void ExplicitMatchTest()
+        public void CombineTest(IEnumerable<bool> inputBooleans, bool expectedResult,
+            StrictIdFilterForTests.EqualValueFunction expectedCall)
+        {
+            var filters = new List<StrictIdFilterForTests>();
```
Would you add a comment in this and other tests like it - or maybe on the StrictIdFilterForTests class that explains what you are tryiing to test. It really takes a bit of mental exercise to figure it out and we would like folks to understand our tests, which are often used as examples.

CharliePoole commented:
Aside from some minor comments, this looks good to me.

stffabi commented:
Yes, I agree with you about clarifying these kind of tests. I think the clearest thing would be to use some kind of mocking framework for tests like these, e.g. Moq. I have taken a look into the source, but couldn't find something like that.

Would you be interested to use a mocking framework for such tests? Would that be in the interest of the project, if yes I would be willing to include e.g. Moq.

CharliePoole commented:
We use NSubstitute where we use mocks, but don't have any now in the 3.0 code. It could be a bit much to introduce just for this fix.

stffabi commented:
Okey, then I will only clarify these tests with comments. The changes will be around tomorrow.

CharliePoole commented:
This looks good to merge for me. @rprouse?

rprouse commented:
Good by me too, thanks for adding the tests. 

FYI, in the future, add a comment after you push updates to a PR letting us know. We get notified by email for comments, but not for code commits, so we sometimes miss updates like this.

Thanks for the contribution.

oznetmaster commented:
Why the use of TestFilter in the foreach statements when previously we have always used ITestFilter?

Not only is it inconsistent, but it throws up warning.
